codeU2.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+def dfs(tree, start):
+
+    """Depth-first search
+    """
+
+    if tree is None:
+        return
+    visited = set()
+    stack = [start]
+    while stack:
+        vertex = stack.pop()
+        if not vertex:
+            continue
+        visited.add(vertex)
+        stack.extend(set(tree[vertex]) - visited)
+    return visited
+
+def dfs_path(tree, start, end, path=None):
+
+    """Returns all full paths from start node
+    to end node, including the end node
+    """
+
+    stack = [(start, [start])]
+    while stack:
+        (vertex, path) = stack.pop()
+        if vertex is None:
+            continue
+        for next in set(tree[vertex]) - set(path):
+            if next == end:
+                yield path + [end]
+            else:
+                stack.append((next, path+[next]))
+
+def get_head(tree):
+
+    """
+    Returns head of the tree.
+
+    Args:
+        tree: dictionary reprenseting binary tree
+        key: is a node, value: (left_child,right_child)
+        for leaves value is ().
+        if right or left child are missing, the value is None
+
+    Returns:
+        head value of the tree
+        None, when tree is empty
+    """
+
+
+    if not tree:
+        return None
+    all_nodes = set()
```

Comment:
Would it make it slightly clearer to rename `all_nodes` to `all_children` and `set(tree.keys())` (below in the final expression) `all_parents`?

I'm specifically looking at the set arithmetic expression below where it is not immediately obvious why `all_nodes` and `tree.keys()` are subtracted as they are and thinking that a little bit of naming might help.